core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+        _w_dp - word given doc and count
+            (list of word ids which represent a doc)
+        _b_dl - boolean given doc and label
+            (boolean of whether a doc has a label)
+        _l_dp - label given doc and count
+            (assigned label of each word instance in a doc)
+        _c_dl - count given doc and label
+            (count of each label per doc)
+        _c_lw - count given label and word
+            (count of labels per word)
+        _c_l - count given label
+            (count of each label type)
+        """
+        # Ensure that results are deterministic
+        # There is nothing special about the value 4
+        numpy.random.seed(seed=4)
+
+        self._alpha = self._DEFAULT_ALPHA
+        self._beta = self._DEFAULT_BETA
+
+    def _get_word_id(self, word):
+        """Returns a word's id if it exists, otherwise assigns
+        a new id to the word and returns it."""
+        if word not in self._word_to_id:
+            self._word_to_id[word] = self._num_words
+            self._num_words += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """Returns a label's id if it exists, otherwise assigns
+        a new id to the label and returns it."""
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_record(self, d):
+        """Given a doc id, return the doc and labels."""
+        return self._w_dp[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """Returns a vector specifying which labels
+        are specified in the training set."""
+        # When we see an empty label list, predicting on all labels
+        # is assumed (versus having to pass all the labels each prediction)
+        if len(labels) == 0:
+            return numpy.ones(self._num_labels)
+        label_vector = numpy.zeros(self._num_labels)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        # Always set default label
+        label_vector[self._label_to_id[self._DEFAULT_LABEL]] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, inc=True):
+        """Update counting matrices given doc/word/label indices.
+        Either increments or decrements."""
+        val = 1 if inc else -1
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._num_docs)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
```

Comment:
II'm curious if a python implementation could actually beat numpy. This really isn't a big optimization winner though since most time is spent in infer. Maybe a fun experiment to try given the time.